### PR TITLE
ci: Add testing on macOS

### LINKED ISSUE #157: Run tests on macOS
For declaring macOS support, we need to run tests on macOS. In this issue, we need to extend test environments in Github Actions with macOS.

Don't forget to add macOS as supported platform in th README.

### PR DESCRIPTION
**Pull request by better0fdead**
Added macOS testing into ci.
- [not needed] Tests
- [Added] Changelog
- [not needed] Documentation 

Closes #157

### CODE REVIEW

File: .github/workflows/testing.yml [hunk 1]
```diff
@@ -184,7 +184,7 @@ jobs:
           rm -f ${ARCHIVE_NAME}
 
       - name: Setup golang for the connector and tests
-        uses: actions/setup-go@v2
+        uses: actions/setup-go@v3
```
Comment (oleg-jukovec) [inline .github/workflows/testing.yml:187]: It is better to bump versions in a separate commit.

File: .github/workflows/testing.yml [hunk 2]
```diff
@@ -240,3 +240,191 @@ jobs:
...
+    #
+    # The main trick is described here:
+    # https://github.com/Dart-Code/Dart-Code/pull/2375
+    if: (github.event_name == 'push') ||
+      (github.event_name == 'pull_request' &&
+        github.event.pull_request.head.repo.full_name != github.repository) ||
+      (github.event_name == 'workflow_dispatch')
+
+    strategy:
+      fail-fast: false
```
Comment (oleg-jukovec) [inline .github/workflows/testing.yml:256]: Why is this needed?

File: .github/workflows/testing.yml [hunk 3]
```diff
@@ -240,3 +240,191 @@ jobs:
...
+          # build either debug or release (RelWithDebInfo), but
+          # we unable to build all releases in debug (see below).
+          #
+          # v3 is to re-verify all Mac OS builds after fix for the
+          # gh-6076 problem (see below).
+          #
+          # v4 added due to inability to clear the cache after v3 prefix.
+          # See https://github.com/github/docs/issues/14145
+          key: ${{ matrix.runs-on }}-${{ matrix.tarantool }}-v4
+        if: matrix.tarantool != 'brew' && matrix.tarantool != 'master'
```
Comment (oleg-jukovec) [inline .github/workflows/testing.yml:301]: I don't see that matrix.tarantool = `matrix` is possible:
```suggestion
        if: matrix.tarantool != 'brew'
```

File: .github/workflows/testing.yml [hunk 4]
```diff
@@ -240,3 +240,191 @@ jobs:
...
+
+      - name: Setup golang for the connector and tests
+        uses: actions/setup-go@v3
+        with:
+          go-version: ${{ matrix.golang }}
+
+      - name: Install test dependencies
+        run:  |
+          brew install luarocks
```
Comment (oleg-jukovec) [inline .github/workflows/testing.yml:397]: Why we need this line? It should be enough just install tarantoolctl/tt.

File: .github/workflows/testing.yml [hunk 5]
```diff
@@ -240,3 +240,191 @@ jobs:
...
+      - name: Run fuzzing tests
+        if: ${{ matrix.fuzzing }}
+        run: |
+          cd "${SRCDIR}"
+          make fuzzing TAGS="go_tarantool_decimal_fuzzing"
+
+      - name: Check workability of benchmark tests
+        run: |
+          cd "${SRCDIR}"
+          make bench-deps bench DURATION=1x COUNT=1
```
Comment (oleg-jukovec) [inline .github/workflows/testing.yml:430]: Could we extract tests in a separate GitHub action/function instead of copying it?

**Review by oleg-jukovec — CHANGES_REQUESTED** [review]
I would like to see a motivation in commit messages for test fixes. In the future it will be difficult to understand why it is necessary.

File: queue/queue_test.go [hunk 6]
```diff
@@ -791,21 +791,29 @@ func TestUtube_Put(t *testing.T) {
...
 	if err := t2.Ack(); err != nil {
+		<-errChan
 		t.Fatalf("Failed to ack task: %s", err)
 	}
 	end := time.Now()
 	if _, ok := <-errChan; ok {
 		t.Fatalf("One of tasks failed")
 	}
-	if math.Abs(float64(end.Sub(start)-2*time.Second)) > float64(200*time.Millisecond) {
+	if math.Abs(float64(end.Sub(start)-2*time.Second)) > float64(700*time.Millisecond) {
```
Comment (DifferentialOrange) [inline queue/queue_test.go:816]: Maths should be more readable

File: .github/workflows/testing.yml [hunk 7]
```diff
@@ -240,3 +240,191 @@ jobs:
...
+        with:
+          repository: tarantool/tarantool
+          ref: ${{ env.T_VERSION }}
+          path: ${{ env.T_TARDIR }}
+          submodules: true
+          # fetch-depth is 1 by default and it is okay for
+          # building from a tag. However we have master in
+          # the version list.
+          fetch-depth: 0
+        if: matrix.tarantool != 'brew' && steps.cache.outputs.cache-hit != 'true'
```
[the remark this answers is not in the document]
Comment (better0fdead) [reply .github/workflows/testing.yml:322]: It would require some workarounds, to work properly.  You can see one https://github.com/tarantool/tt/issues/282. Also it would probably require more manipulations, because for now tt builds tarantool that works properly inside tt's environment. We've discussed it and probably soon will make a solution for it.

[the remark this answers is not in the document]
Comment (oleg-jukovec) [reply .github/workflows/testing.yml:321]: I just did it recently in this repository for Linux:

https://github.com/tarantool/go-tarantool/blob/f300bd920138a16dfb881428a32650b34e1362f5/.github/workflows/testing.yml#L64-L89

It looks simpler than the code we have now in the pool request. It seems to me that it is better to use `tt` if there are no additional issue on the MacOs.

File: .github/workflows/testing.yml [hunk 8]
```diff
@@ -240,3 +240,191 @@ jobs:
...
+          cd "${T_TARDIR}"
+          # These steps fix the problem with tarantool build described in
+          # https://github.com/tarantool/tarantool/issues/6576
+          git show 11e87877df9001a4972019328592d79d55d1bb01 | patch -p1 -f
+        if: matrix.tarantool != 'brew' && matrix.tarantool != 'master' && steps.cache.outputs.cache-hit != 'true'
+
+      - name: Build tarantool ${{ env.T_VERSION }} from sources
+        run: |
+          cd "${T_TARDIR}"
+          # Set RelWithDebInfo just to disable -Werror.
```
[the remark this answers is not in the document]
Comment (better0fdead) [reply .github/workflows/testing.yml:326]: no, it is copied from tarantool/smtp

File: connection_pool/connection_pool_test.go [hunk 9]
```diff
@@ -2192,7 +2196,7 @@ func TestConnectionPool_NewWatcher_update(t *testing.T) {
 			} else {
 				testMap[addr] = 1
 			}
-		case <-time.After(time.Second):
+		case <-time.After(poolOpts.CheckTimeout * 2):
```
[the remark this answers is not in the document]
Comment (better0fdead) [reply connection_pool/connection_pool_test.go:2199]: The reason is that by default it updates the connects once every 1 second.
And it doesn't have time to update for some reason in the case of a test in the same second.
That's why we set the update to half the time.
Couldn't reproduce locally, seems like very slow GitHub runners for macOs.

File: connection_pool/connection_pool_test.go [hunk 10]
```diff
@@ -2164,7 +2164,11 @@ func TestConnectionPool_NewWatcher_update(t *testing.T) {
 	err := test_helpers.SetClusterRO(servers, opts, roles)
 	require.Nilf(t, err, "fail to set roles for cluster")
 
-	pool, err := connection_pool.Connect(servers, opts)
+	poolOpts := connection_pool.OptsPool{
+		CheckTimeout: 500 * time.Millisecond,
```
[the remark this answers is not in the document]
Comment (better0fdead) [reply connection_pool/connection_pool_test.go:2168]: The reason is that by default it updates the connects once every 1 second.
And it doesn't have time to update for some reason in the case of a test in the same second.
That's why we set the update to half the time.
Couldn't reproduce locally, seems like very slow GitHub runners for macOs.

File: .github/workflows/testing.yml (hunk 2 above)
Comment (better0fdead) [reply .github/workflows/testing.yml:256]: It is needed because w/o it if for example macOS-11 1.10 tarantool test fails other jobs are also cancelled.

File: .github/workflows/testing.yml [hunk 11]
```diff
@@ -240,3 +240,191 @@ jobs:
...
+          # debug tarantool builds. It is desirable to have all
+          # build either debug or release (RelWithDebInfo), but
+          # we unable to build all releases in debug (see below).
+          #
+          # v3 is to re-verify all Mac OS builds after fix for the
+          # gh-6076 problem (see below).
+          #
+          # v4 added due to inability to clear the cache after v3 prefix.
+          # See https://github.com/github/docs/issues/14145
+          key: ${{ matrix.runs-on }}-${{ matrix.tarantool }}-v4
```
[the remark this answers is not in the document]
Comment (better0fdead) [reply .github/workflows/testing.yml]: removed

File: .github/workflows/testing.yml (hunk 7 above)
[the remark this answers is not in the document]
Comment (better0fdead) [reply .github/workflows/testing.yml:311]: after discussion in voice chat decided to keep my variant. After publishing tt to home-brew will be changed.

File: .github/workflows/testing.yml [hunk 12]
```diff
@@ -240,3 +240,191 @@ jobs:
...
+      SRCDIR: ${{ format('{0}/{1}', github.workspace, github.repository) }}
+
+    runs-on: ${{ matrix.runs-on }}
+    steps:
+      - name: Clone the connector
+        uses: actions/checkout@v3
+        with:
+          path: ${{ env.SRCDIR }}
+
+      - name: Cache built tarantool ${{ env.T_VERSION }}
```
[the remark this answers is not in the document]
Comment (better0fdead) [reply .github/workflows/testing.yml]: renamed

File: .github/workflows/testing.yml [hunk 13]
```diff
@@ -240,3 +240,191 @@ jobs:
...
+      fail-fast: false
+      matrix:
+        golang:
+          - 1.13
+        runs-on:
+          - macos-11
+          - macos-12
+        tarantool:
+          - brew
+          - 1.10.10
```
[the remark this answers is not in the document]
Comment (better0fdead) [reply .github/workflows/testing.yml]: changed

File: .github/workflows/testing.yml (hunk 5 above)
Comment (better0fdead) [reply .github/workflows/testing.yml:411]: I don't think so, because we make tests differently for linux/macOs.
ee:
```
      - name: Run regression tests with call_17
        run: |
          source tarantool-enterprise/env.sh
          make test TAGS="go_tarantool_call_17"
          make testrace TAGS="go_tarantool_call_17"
        env:
          TEST_TNT_SSL: ${{matrix.ssl}}
```

ce:
```
      - name: Run regression tests with call_17
        run: |
          make test TAGS="go_tarantool_call_17"
```

ee:
```
      - name: Run regression tests with call_17
        run: |
          cd "${SRCDIR}"
          make test TAGS="go_tarantool_call_17"
```

Comment (oleg-jukovec) [reply .github/workflows/testing.yml:411]: Why are there no `testrace` on macOS?

File: queue/queue_test.go (hunk 6 above)
Comment (better0fdead) [reply queue/queue_test.go]: changed

File: .github/workflows/testing.yml (hunk 5 above)
Comment (better0fdead) [reply .github/workflows/testing.yml]: added

File: .github/workflows/testing.yml (hunk 4 above)
Comment (better0fdead) [reply .github/workflows/testing.yml:384]: For me it is easier to install luarocks than tarantoolctl/tt since it is in homebrew. When tt will be published to homebrew it can be changed with install of tarantool step.

File: .github/workflows/testing.yml (hunk 1 above)
Comment (better0fdead) [reply .github/workflows/testing.yml:187]: changed

**Review by oleg-jukovec — CHANGES_REQUESTED** [review]
I see a flaky test on macOS:
```
2023-03-16T14:18:05.7459090Z --- FAIL: TestReconnect (0.62s)
2023-03-16T14:18:05.7567240Z     multi_test.go:108: conn has incorrect addr: 127.0.0.1:3013 after disconnect server1
```

I think you need to make a change similar too:

https://github.com/tarantool/go-tarantool/commit/b0956f307e177134a247277ba014f5c06e997e01

1. Stop the tarantool instance instead of just conn.Close() to avoid auto-reconnects.
2. Ensure that the connection is gone from connection pool.
3. Restart the instance.
4. Ensure that the connection is available for requests.

File: multi/multi_test.go [hunk 14]
```diff
@@ -92,28 +92,51 @@ func TestConnSuccessfully(t *testing.T) {
...
+	test_helpers.StopTarantoolWithCleanup(instances[0])
 
-	conn, _ := multiConn.getConnectionFromPool(server1)
-	conn.Close()
+	for i := 0; i < sleepCnt; i++ {
+		_, ok := multiConn.getConnectionFromPool(servers[0])
+		if !ok {
+			break
+		}
+		multiConn.getState()
```
Comment (oleg-jukovec) [inline multi/multi_test.go:110]:
```suggestion
```

**Review by oleg-jukovec — APPROVED** [review]
